### PR TITLE
docs(concepts/auditing): Add description of BSR

### PR DESCRIPTION
**Pull request by stasryzhov**
This PR adds screenshots of various BSR directories uploaded to S3

### CODE REVIEW

**Comment by stasryzhov** [comment]
If you find the screenshots hard to read, let's decide if we want to show the files in the red box only

File: website/content/docs/concepts/auditing.mdx [hunk 1]
```diff
@@ -72,9 +72,15 @@ It contains all the data transmitted between a user and a target during a single
 
 Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This top level directory contains session summary information and subdirectories for connections.
 
+![BSR top-level directory](/img/session-recording-bsr-top-level.png)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:75]:
```suggestion
[![Screenshot of a BSR top-level directory](/img/session-recording-bsr-top-level.png)](/img/session-recording-bsr-top-level.png)
```

File: website/content/docs/concepts/auditing.mdx [hunk 2]
```diff
@@ -72,9 +72,15 @@ It contains all the data transmitted between a user and a target during a single
...
+
 A BSR connections directory contains a summary of connections, as well as inbound and outbound requests.
+
+![BSR connection-level directory](/img/session-recording-bsr-connection-level.png)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:79]:
```suggestion
[![Screenshot of a BSR connection-level directory](/img/session-recording-bsr-connection-level.png)](/img/session-recording-bsr-connection-level.png)
```

File: website/content/docs/concepts/auditing.mdx [hunk 3]
```diff
@@ -72,9 +72,15 @@ It contains all the data transmitted between a user and a target during a single
...
+
 If you use a multiplexed protocol, there are subdirectories for the channels.
 
+![BSR channel-level directory](/img/session-recording-bsr-channel-level.png)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:83]:
```suggestion
[![Screenshot of a BSR channel-level directory](/img/session-recording-bsr-channel-level.png)](/img/session-recording-bsr-channel-level.png)
```

**Review by Dan-Heath — COMMENTED** [review]
Thank you for taking this on Stan! I think it is a nice addition to this topic. IMO the screen shots look good. But I was thinking we could wrap them in a link to themselves, so if the user clicks it, it opens in full screen. What do you think?

**Review by Dan-Heath — APPROVED** [review]
LGTM, thank you!!

**Comment by stasryzhov** [comment]
As discussed, the screenshots are replaced with text. I've added only one example without breaking it into three parts as it was done with screenshots. The text form feels and reads more natural to me this way

**Review by ddebko — APPROVED** [review]
This looks great

File: website/content/docs/concepts/auditing.mdx [hunk 4]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+## Validating the Integrity of Session Recordings
 
 BSR directories are validated based on the contents in the directory.
-Each BSR directory contains a SHA256SUMS and SHA256SUMS.sign file that can be used to cryptographically verify the BSR directory's contents.
-The SHA256SUMS file contains rows of file names paired with a checksum for the file contents.
-The SHA256SUMS.sign is a copy of the SHA256SUMS file, signed with the BSR's private key.
+Boundary cryptographically verifies each individual Boundary Session Recording (BSR) file. The keys used for verifying all
+Boundary Session Recording files are written to storage and are wrapped by your configured KMS. Each Session Recording will
+have its own individual key. The following keys will be generated when a Session Recording is authorized:
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:138]:
```suggestion
Boundary Session Recording files are written to storage and are wrapped by the KMS you configured. Each session recording has its own individual key. Boundary generates the following keys when a session recording is authorized:
```

File: website/content/docs/concepts/auditing.mdx [hunk 5]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
 
+- BSR Key: AES-GCM used for signing the public key.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:140]:
```suggestion
- BSR key: An AES-GCM that is used for signing the public key.
```

File: website/content/docs/concepts/auditing.mdx [hunk 6]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
 
+- Private/Public key pair: ed25519 key pair.  Private key is used to sign BSR files: initially the public key and the BSR checksums file.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:142]:
```suggestion
- Private/public key pair: ed25519 key pair.  A private key is used to sign the public key and the BSR checksums file. 
```

File: website/content/docs/concepts/auditing.mdx [hunk 7]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
-The SHA256SUMS.sign is a copy of the SHA256SUMS file, signed with the BSR's private key.
+Boundary cryptographically verifies each individual Boundary Session Recording (BSR) file. The keys used for verifying all
+Boundary Session Recording files are written to storage and are wrapped by your configured KMS. Each Session Recording will
+have its own individual key. The following keys will be generated when a Session Recording is authorized:
 
+- BSR Key: AES-GCM used for signing the public key.
 
+- Private/Public key pair: ed25519 key pair.  Private key is used to sign BSR files: initially the public key and the BSR checksums file. 
+
+The BSR’s key and the derived key-pair will be written to storage beside all the other BSR files.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:144]:
```suggestion
The BSR’s key and the derived key-pair are written to storage with all the other BSR files.
```

File: website/content/docs/concepts/auditing.mdx [hunk 8]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+Encrypting the BSR key with an external KMS means that Boundary is not responsible for the longevity of the keys.
+The Boundary admin can always use that external KMS they control to decrypt BSR keys.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:146]:
```suggestion
The Boundary admin can always use that external KMS to decrypt BSR keys. 
```

File: website/content/docs/concepts/auditing.mdx [hunk 9]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
 
+- BSR Key: AES-GCM used for signing the public key.
 
+- Private/Public key pair: ed25519 key pair.  Private key is used to sign BSR files: initially the public key and the BSR checksums file. 
+
+The BSR’s key and the derived key-pair will be written to storage beside all the other BSR files.
+Encrypting the BSR key with an external KMS means that Boundary is not responsible for the longevity of the keys.
+The Boundary admin can always use that external KMS they control to decrypt BSR keys. 
+A BSR’s key will be encrypted using the `go-kms-wrapping` package and therefore the encrypted BlobInfo will include the metadata required to identify the key-version used during encryption,
+so if the wrapper is reinitialized properly; the keys can be unwrapped even if the key has been rotated.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:148]:
```suggestion
A BSR’s key is encrypted using the `go-kms-wrapping` package, and therefore the encrypted BlobInfo includes the metadata required to identify the key-version used during encryption.
So if the wrapper is reinitialized properly, you can unwrap the keys even if the key has been rotated.  
```

File: website/content/docs/concepts/auditing.mdx [hunk 10]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+- BSR Key: AES-GCM used for signing the public key.
 
+- Private/Public key pair: ed25519 key pair.  Private key is used to sign BSR files: initially the public key and the BSR checksums file. 
+
+The BSR’s key and the derived key-pair will be written to storage beside all the other BSR files.
+Encrypting the BSR key with an external KMS means that Boundary is not responsible for the longevity of the keys.
+The Boundary admin can always use that external KMS they control to decrypt BSR keys. 
+A BSR’s key will be encrypted using the `go-kms-wrapping` package and therefore the encrypted BlobInfo will include the metadata required to identify the key-version used during encryption,
+so if the wrapper is reinitialized properly; the keys can be unwrapped even if the key has been rotated.  
+Users can use `go-kms-wrapping` independently from Boundary to unwrap the BSR keys as needed.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:149]:
```suggestion
You can use `go-kms-wrapping` independently from Boundary to unwrap the BSR keys as needed.
```

File: website/content/docs/concepts/auditing.mdx [hunk 11]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+Each BSR directory contains a SHA256SUM and SHA256SUM.sign file that can be used to cryptographically verify the BSR directory's contents.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:151]:
```suggestion
Each BSR directory contains a SHA256SUM and SHA256SUM.sign file that you can use to cryptographically verify the BSR directory's contents.
```

File: website/content/docs/concepts/auditing.mdx [hunk 12]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+The SHA256SUM file contains rows of file names paired with a checksum for the file contents.
+The SHA256SUM.sign is a copy of the SHA256SUM file, signed with the BSR's private key.
+An example of the SHA256SUM file:
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:154]:
```suggestion
Refer to the following example of a SHA256SUM file:
```

File: website/content/docs/concepts/auditing.mdx [hunk 13]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+   ad76483e7cf3e65391a3a1d0b86a3ad436333ee225bea042b13900abc188b226  session-recording-summary.json
+   ```
+
+Workers will use these keys to sign BSR data as follows:
+
+- The BSR key will be used to sign the public key using `go-kms-wrapping` package HmacSha256(...)
+
+- The private key will be used to sign the public key using `go-kms-wrapping` package ed25519.Sign(...)
+
+- The private key will be used to sign the BSR SHA256SUM file using `go-kms-wrapping` package ed25519.Sign(...)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:173]:
```suggestion
Workers use these keys to sign BSR data as follows:

- The BSR key is used to sign the public key using `go-kms-wrapping` package HmacSha256(...)

- The private key is used to sign the public key using `go-kms-wrapping` package ed25519.Sign(...)

- The private key is used to sign the BSR SHA256SUM file using `go-kms-wrapping` package ed25519.Sign(...)
```

File: website/content/docs/concepts/auditing.mdx [hunk 14]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+Workers will use these keys to verify BSR data as follows:
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:175]:
```suggestion
Workers use these keys to verify BSR data as follows:
```

File: website/content/docs/concepts/auditing.mdx [hunk 15]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+- The BSR key or the private key can be used to verify the public key using `go-kms-wrapping` HmacSha256(...)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:177]:
```suggestion
- You can use the BSR key or the private key to verify the public key using `go-kms-wrapping` HmacSha256(...)
```

File: website/content/docs/concepts/auditing.mdx [hunk 16]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+    ├── pubKeyBsrSignature.sign
+    ├── pubKeySelfSignature.sign
+    ├── session-meta.json
+    ├── session-recording-summary.json
+    ├── session-recording.meta
+    ├── wrappedBsrKey
+    └── wrappedPrivKey
+```
+
+## Validating the Integrity of Session Recordings
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:133]:
```suggestion
## Validate the integrity of session recordings
```

File: website/content/docs/concepts/auditing.mdx [hunk 17]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+- The private key will be used to sign the BSR SHA256SUM file using `go-kms-wrapping` package ed25519.Sign(...)
+
+Workers will use these keys to verify BSR data as follows:
+
+- The BSR key or the private key can be used to verify the public key using `go-kms-wrapping` HmacSha256(...)
+
+- Once verified, the public key can be used to verify the BSR SHA256SUM file using `go-kms-wrapping` ed25519.Sign(...)
+
+## Validating the Data Integrity in the External Object Store
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:181]:
```suggestion
## Validate the data integrity in the external object store
```

File: website/content/docs/concepts/auditing.mdx [hunk 18]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+- The private key will be used to sign the public key using `go-kms-wrapping` package ed25519.Sign(...)
+
+- The private key will be used to sign the BSR SHA256SUM file using `go-kms-wrapping` package ed25519.Sign(...)
+
+Workers will use these keys to verify BSR data as follows:
+
+- The BSR key or the private key can be used to verify the public key using `go-kms-wrapping` HmacSha256(...)
+
+- Once verified, the public key can be used to verify the BSR SHA256SUM file using `go-kms-wrapping` ed25519.Sign(...)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:179]:
```suggestion
- Once verified, you can use the public key to verify the BSR SHA256SUM file using `go-kms-wrapping` ed25519.Sign(...)
```

File: website/content/docs/concepts/auditing.mdx [hunk 19]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+## Validating the Data Integrity in the External Object Store
+
+When a Boundary worker is uploading a BSR file to AWS S3 through the Boundary AWS Plugin, the worker calculates the SHA256 checksum of the
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:183]:
```suggestion
When a Boundary worker uploads a BSR file to AWS S3 through the Boundary AWS plugin, the worker calculates the SHA256 checksum of the
```

File: website/content/docs/concepts/auditing.mdx [hunk 20]
```diff
@@ -75,12 +75,114 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+contents of the BSR file and attaches this information to the object that is uploaded to S3. The stored SHA256 checksum value can now be used
+to validate the contents of the BSR file when downloading it from AWS S3 through the Boundary AWS Plugin. This process ensures that
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:185]:
```suggestion
contents of the BSR file and attaches this information to the object that is uploaded to S3. You can now use the stored SHA256 checksum value to validate the contents of the BSR file when you download it from AWS S3 using the Boundary AWS plugin. This process ensures that
```

**Review by Dan-Heath — CHANGES_REQUESTED** [review]
Thank you for taking this on! I added some suggestions for style and tone, but the content in the two new validation sections is great. I am wondering if they would make more sense as a new topic (or topics), though. This topic is under **Concepts** in the table of contents, but the new sections are more like how-to topics where the user takes action. Would they make more sense under:

Configuration > Session recordings

**Comment by AdamBouhmad** [comment]
This looks great. Do we elaborate anywhere on the intent behind each of the files outputted in the directories? I think that would be super helpful context for a user versus having to learn as they go

**Comment by stasryzhov** [comment]
> This looks great. Do we elaborate anywhere on the intent behind each of the files outputted in the directories? I think that would be super helpful context for a user versus having to learn as they go

Working on it together with @ddebko at the moment 👍

File: website/content/docs/concepts/auditing.mdx (hunk 20 above)
Comment (ddebko) [reply website/content/docs/concepts/auditing.mdx:380]: The user doesn't have access to the SHA256 checksum value when using the Boundary AWS plugin to download a BSR file. I think if I change `You can now use the stored SHA256 checksum value` to `The worker can now use the stored...` 

Users can access the SHA256 checksum value of each BSR file if they access it directly from S3 and not through Boundary.

**Comment by AdamBouhmad** [comment]
Amazing contribution @stasryzhov & @ddebko. @stellarsquall / @Dan-Heath curious if you think there should be any updates to the length, or if the content is in the right location. Otherwise this looks like a fantastic first pass imo

**Review by AdamBouhmad — APPROVED** [review]

File: website/content/docs/concepts/auditing.mdx (hunk 20 above)
Comment (ddebko) [reply website/content/docs/concepts/auditing.mdx:380]: Does this make more sense?
```
When a Boundary worker uploads a BSR file to AWS S3 through the Boundary AWS plugin, the plugin calculates the SHA256 checksum of the
contents of the BSR file and attaches this information to the object that is uploaded to S3. The SHA256 checksum value attached to the
S3 object is returned to the Boundary worker. The Boundary worker will calculate the SHA256 checksum value of the BSR file's content from
local disk and compare it to the plugin value. This process ensures that no tampering of BSR files occurs between the worker, plugin, and S3.
The SHA256 checksum value generated by the plugin is not a part of the BSR file structure and should not be confused with how Boundary
cryptographically verifies BSR directory's contents.
```

File: website/content/docs/concepts/auditing.mdx [hunk 21]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+`session-recording.json` file example:
+
+```
+id: sr_iNCdGSREeX
+protocol: BSSH
+connection: cr_3bB78W53Y9.connection
+```
+
+`SHA256SUM` and `SHA256SUM.sig` files are to be used for cryptographically verifying the contents of this directory.
+For more information on `*.sign`, `bsrKey.pub`, `wrappedBsrKey`, and `wrappedPrivKey` files, refer to the [Validating the Integrity of Session Recordings](#validating-the-integrity-of-session-recordings) paragraph.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:239]:
```suggestion
`SHA256SUM` and `SHA256SUM.sig` files are used for cryptographically verifying the contents of this directory.
For more information on `*.sign`, `bsrKey.pub`, `wrappedBsrKey`, and `wrappedPrivKey` files, refer to [Validating the integrity of session recordings](#validating-the-integrity-of-session-recordings).
```

File: website/content/docs/concepts/auditing.mdx [hunk 22]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+  "StartTime": "2023-07-13T20:21:49.164105381Z",
+  "EndTime": "2023-07-13T20:22:37.241911112Z",
+  "BytesUp": 125,
+  "BytesDown": 748,
+  "Errors": null
+}
+```
+
+`*.data` files are binary files containing all data transmitted during a session.
+`SHA256SUM` and `SHA256SUM.sig` files are to be used for cryptographically verifying the contents of this directory.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:278]:
```suggestion
`SHA256SUM` and `SHA256SUM.sig` files are used for cryptographically verifying the contents of this directory.
```

File: website/content/docs/concepts/auditing.mdx [hunk 23]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+  },
+  "SessionProgram": "shell",
+  "SubsystemName": "",
+  "ExecProgram": "",
+  "FileTransferDirection": "not applicable"
+}
+```
+
+`*.data` files are binary files containing all data transmitted during a session.
+`SHA256SUM` and `SHA256SUM.sig` files are to be used for cryptographically verifying the contents of this directory.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:326]:
```suggestion
`SHA256SUM` and `SHA256SUM.sig` files are used for cryptographically verifying the contents of this directory.
```

File: website/content/docs/concepts/auditing.mdx [hunk 24]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+## Validate the integrity of session recordings
 
 BSR directories are validated based on the contents in the directory.
-Each BSR directory contains a SHA256SUMS and SHA256SUMS.sign file that can be used to cryptographically verify the BSR directory's contents.
-The SHA256SUMS file contains rows of file names paired with a checksum for the file contents.
-The SHA256SUMS.sign is a copy of the SHA256SUMS file, signed with the BSR's private key.
+Boundary cryptographically verifies each individual Boundary Session Recording (BSR) file. The keys used for verifying all
+Boundary Session Recording files are written to storage and are wrapped by the KMS you configured. Each session recording
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:332]:
```suggestion
Boundary Session Recording files are written to storage and wrapped by the KMS you configured. Each session recording 
```

File: website/content/docs/concepts/auditing.mdx [hunk 25]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
-Each BSR directory contains a SHA256SUMS and SHA256SUMS.sign file that can be used to cryptographically verify the BSR directory's contents.
-The SHA256SUMS file contains rows of file names paired with a checksum for the file contents.
-The SHA256SUMS.sign is a copy of the SHA256SUMS file, signed with the BSR's private key.
+Boundary cryptographically verifies each individual Boundary Session Recording (BSR) file. The keys used for verifying all
+Boundary Session Recording files are written to storage and are wrapped by the KMS you configured. Each session recording 
+has its own individual key. Boundary generates the following keys when a session recording is authorized:
+
+- BSR key is a plaintext AES-GCM key. (This is not uploaded to the external object store)
+
+- Private and Public key pair is a ed25519 key pair. (The key pair is not uploaded to the external object store)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:337]:
```suggestion
-  The BSR key is a plaintext AES-GCM key. It is not uploaded to the external object store.

- The private and public key pair is a ed25519 key pair. The key pair is not uploaded to the external object store.
```

File: website/content/docs/concepts/auditing.mdx [hunk 26]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+- `wrappedBsrKey` is the BSR key wrapped by the external KMS AES-GCM key configured by you.
+
+- `wrappedPrivKey` is the Private ed25519 key wrapped by the external KMS AES-GCM key configured by you.
+
+- `pubKeySelfSignature.sign` is a self-signature of the plaintext Public ed25519 key created with its Private key.
+
+- `pubKeyBsrSignature.sign` is a signature of the plaintext Public ed25519 key created with the BSR key.
+
+- `SHA256SUM.sig` is a signature of the plaintext `SHA256SUM` file created with the Private key.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:351]:
```suggestion
- `bsrKey.pub` is the public ed25519 key.

- `wrappedBsrKey` is the BSR key wrapped by the external KMS AES-GCM key that you configure.

- `wrappedPrivKey` is the private ed25519 key wrapped by the external KMS AES-GCM key that you configure.

- `pubKeySelfSignature.sign` is a self-signature of the plaintext public ed25519 key created with its private key.

- `pubKeyBsrSignature.sign` is a signature of the plaintext public ed25519 key created with the BSR key.

- `SHA256SUM.sig` is a signature of the plaintext `SHA256SUM` file created with the private key.
```

File: website/content/docs/concepts/auditing.mdx [hunk 27]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+Encrypting the BSR key with an external KMS means that Boundary is not responsible for the longevity of the keys.
+The Boundary admin can always use that external KMS to unwrap the `wrappedBsrKey` and `wrappedPrivKey`.
+A BSR’s key is encrypted using the `go-kms-wrapping` package, and therefore the encrypted BlobInfo will includes the metadata required to identify the key-version used during encryption.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:355]:
```suggestion
A BSR’s key is encrypted using the `go-kms-wrapping` package, and therefore the encrypted BlobInfo includes the metadata required to identify the key-version used during encryption.
```

File: website/content/docs/concepts/auditing.mdx [hunk 28]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+   a5a91b1b52fb53c4bab661b2e5846bb2a836f050e3d745e436078871914a0bc2  wrappedPrivKey
+   1ca281852ec0d447b94708f28a51b562d47b84affdba25e13a97b0fbd9126424  pubKeyBsrSignature.sign
+   7b5e18e5930bb4cce12a3f203328d9065cae29f26aba3963bb5faece2cf97231  pubKeySelfSignature.sign
+   dc7c6b1316624c7c486a22bab157f947df92b9f2ce4a72469b1f335399a043d5  bsrKey.pub
+   4d3966c458f4e5d67f9ac70b804540b927c718965267c3f36526bf0b18c40ad9  session-meta.json
+   6fec2173d331828677fb5e77fc19168daad3c5f0e82517a82e5701e6c2bdcbe1  session-recording.meta
+   ad76483e7cf3e65391a3a1d0b86a3ad436333ee225bea042b13900abc188b226  session-recording-summary.json
+   ```
+
+You can follow these steps to validate a session recording:
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:374]:
```suggestion
Follow these steps to validate a session recording:
```

File: website/content/docs/concepts/auditing.mdx [hunk 29]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+
+1. Unwrap `wrappedBsrKey` using the external KMS configured by you to retrieve the BSR key.
+2. Unwrap `wrappedPrivKey` using the external KMS configured by you to retrieve the Private key.
+3. Use the BSR key or the Private key to verify the `bsrKey.pub` key using `go-kms-wrapping` HmacSha256(...)
+4. Once verified, you can use the `bsrKey.pub` key to verify the BSR SHA256SUM file using `go-kms-wrapping` ed25519.Sign(...)
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:379]:
```suggestion
1. Unwrap `wrappedBsrKey` using the external KMS you configured to retrieve the BSR key.
2. Unwrap `wrappedPrivKey` using the external KMS you configured to retrieve the private key.
3. Use the BSR key or the private key to verify the `bsrKey.pub` key using `go-kms-wrapping` HmacSha256(...)
4. When the key is verified, use the `bsrKey.pub` key to verify the BSR SHA256SUM file using `go-kms-wrapping` ed25519.Sign(...)
```

File: website/content/docs/concepts/auditing.mdx [hunk 30]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
 
+## Validate the data integrity in the external object store
 
+When a Boundary worker uploads a BSR file to AWS S3 through the Boundary AWS plugin, the plugin calculates the SHA256 checksum of the
+contents of the BSR file and attaches this information to the object that is uploaded to S3. The SHA256 checksum value attached to the
+S3 object is returned to the Boundary worker. The Boundary worker will calculate the SHA256 checksum value of the BSR file's content from
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:385]:
```suggestion
S3 object is returned to the Boundary worker. The Boundary worker calculates the SHA256 checksum value of the BSR file's content from
```

File: website/content/docs/concepts/auditing.mdx [hunk 31]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+local disk and compare it to the plugin value. This process ensures that no tampering of BSR files occurs between the worker, plugin, and S3.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:386]:
```suggestion
local disk and compares it to the plugin value. This process ensures that no tampering of BSR files occurs between the worker, plugin, and S3.
```

File: website/content/docs/concepts/auditing.mdx [hunk 32]
```diff
@@ -75,12 +75,316 @@ Boundary creates the top-level directory of the BSR as `<sessionID>.bsr`. This t
...
+The SHA256 checksum value generated by the plugin is not a part of the BSR file structure and should not be confused with how Boundary
+cryptographically verifies BSR directory's contents.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/auditing.mdx:388]:
```suggestion
cryptographically verifies the BSR directory's contents.
```

File: website/content/docs/concepts/auditing.mdx (hunk 20 above)
Comment (Dan-Heath) [reply website/content/docs/concepts/auditing.mdx]: Yes! I think that's much more clear. 👍

**Review by Dan-Heath — CHANGES_REQUESTED** [review]
This is looking really good. Great job all! I just had a few minor suggestions. Please let me know if you have any questions.

I do think the 2 Validating sections should be split off into their own topics, but would be fine with merging this and doing that part later.

**Comment by Dan-Heath** [comment]
> Amazing contribution @stasryzhov & @ddebko. @stellarsquall / @Dan-Heath curious if you think there should be any updates to the length, or if the content is in the right location. Otherwise this looks like a fantastic first pass imo

Yeah, I think this is looking really good! I do think the 2 validation sections should be moved to standalone topics. To me this **Auditing** topic is a more conceptual/higher level take on the benefits of using Boundary - likely something for a stakeholder or decision maker. I'm worried those validation procedures are buried at the bottom and they might be hard to find. I would rather put them somewhere on the TOC, so they are easier to find. They might work under **Configuration** > **Session recordings** or we could nest them under **Operations** > **Session recordings**.

But I am also fine with merging this now, and moving those 2 sections later, if that's how the team wants to proceed. We have some upcoming information architecture changes planned and we can do it as part of that.

**Review by Dan-Heath — APPROVED** [review]
This looks great, thank you!!

**Pull request merged**